airflow/contrib/hooks/aws_athena_hook.py

```diff
@@ -0,0 +1,147 @@
+# specific language governing permissions and limitations
+# under the License.
+
+from time import sleep
+from airflow.contrib.hooks.aws_hook import AwsHook
+from uuid import uuid4
+
+INTERMEDIATE_STATES = ('QUEUED', 'RUNNING',)
+FINAL_STATES = ('SUCCEEDED', 'FAILED', 'CANCELLED',)
+
+
+class AWSAthenaHook(AwsHook):
+
+    def __init__(self, query, database, region_name, output_location, aws_conn_id=None, *args, **kwargs):
+        self.query = query
+        self.database = database
+        self.region_name = region_name
+        self.outputLocation = output_location
+        self.aws_conn_id = aws_conn_id
+        if 'client_request_token' in kwargs:  # This ensures idempotency of query execution
+            self.client_request_token = kwargs['client_request_token']
+        else:
+            self.client_request_token = str(uuid4())
+        if 'query_execution_context' in kwargs:
+            self.query_execution_context = kwargs['query_execution_context']
+        else:
+            self.query_execution_context = {}
+        if 'result_configuration' in kwargs:
+            self.result_configuration = kwargs['result_configuration']
+        else:
+            self.result_configuration = {}
+        if 'sleep_time' in kwargs:
+            self.sleep_time = kwargs['sleep_time']
+        else:
+            self.sleep_time = 5  # Use this to poll query status in athena
+        self.query_execution_context['Database'] = self.database
+        self.result_configuration['OutputLocation'] = self.outputLocation
+        super(AWSAthenaHook, self).__init__(self.aws_conn_id, verify=None)
+        self.conn = self.get_conn()
+
+    def _check_query_status(self):
+        if self._queryExecutionId is None:
+            return None
+        return self.conn.get_query_execution(QueryExecutionId=self._queryExecutionId)
+
+    def _get_query_results(self):
+        if self._queryExecutionId is None:
+            return None
+        state = self.check_query_status()
+        if state is None or state in INTERMEDIATE_STATES:
+            self.log.error("Invalid Query state")
+            return None
+        elif state in INTERMEDIATE_STATES or state == 'CANCELLED':
+            self.log.info("Query is in {state} state. Cannot fetch results".format(state=state))
+            return None
+        return self.conn.get_query_results(QueryExecutionId=self._queryExecutionId)
+
+    def _stop_query(self):
+        if self._queryExecutionId is None:
+            return None
+        return self.conn.stop_query_execution(QueryExecutionId=self._queryExecutionId)
+
+    def _poll_query_status(self, sleep_time=None, max_tries=None):
+        try_number = 1
+        if sleep_time is None:
+            sleep_time = self.sleep_time
+        if max_tries is None:  # Retries forever until query reaches final state
+            retry_condition = True
+        else:
+            retry_condition = try_number <= max_tries
```

Comment:
This doesn't work as you want/intend. This is evaluated one to true or false (almost always True) and then loops forever.

Reply:
@ashb This **retry_condition** variable will be evaluated at the end of each loop if max_retries is provided. This will run forever only when max_retries is not provided or else loop is gonna break after number of tries cross **max_tries**

Reply:
Ah I missed that, sorry

Instead of evaluating the condition at the start (which would only fail if max_tries is 0) how about this:

```
while True:
    # checks

   if try_number > max_tries:
     break
     # Or raise exception etc.
```

Reply:
Also: if it gets to the last try and it's still running it just silently fails I think? that should be an error condition.

Reply:
Thats true @ashb. If query runs even after final polling completes, we cant do anything except to exit the function